File: static/js/contextual_help.js
// ملف JavaScript للمساعدة السياقية
// Contextual Help JavaScript

class ContextualHelp {
    constructor() {
        this.helpData = {};
        this.currentContext = null;
        this.init();
    }
    
    init() {
        console.log('Initializing Contextual Help System...');
        
        // تحميل بيانات المساعدة
        this.loadHelpData();
        
        // إضافة أيقونات المساعدة
        setTimeout(() => this.addHelpIcons(), 500);
        
        // إضافة زر المساعدة العائم
        this.addFloatingButton();
        
        // معالجة الأحداث
        this.bindEvents();
        
        console.log('Contextual Help System initialized');
    }
    
    loadHelpData() {
        // تحميل من API
        fetch('/api/help/tooltips/')
            .then(response => response.json())
            .then(data => {
                this.helpData = data;
                console.log('Help data loaded:', Object.keys(data).length, 'items');
            })
            .catch(error => {
                console.warn('Could not load help data from API, using defaults');
                // استخدام بيانات افتراضية
                this.useDefaultHelpData();
            });
    }
    
    useDefaultHelpData() {
        // بيانات مساعدة افتراضية
        this.helpData = {
            'customer': {
                title: 'اختيار العميل',
                content: 'اختر العميل من القائمة أو ابحث عنه. يمكنك إضافة عميل جديد بالضغط على زر "+".',
                tips: ['استخدم Ctrl+F للبحث السريع', 'العملاء الأكثر تعاملاً يظهرون في الأعلى']
            },
            'discount': {
                title: 'الخصم',
                content: 'أدخل قيمة الخصم بالريال أو النسبة المئوية.',
                tips: ['اضغط F8 لآلة حاسبة الخصم']
            }
        };
    }
    
    addHelpIcons() {
        // البحث عن جميع الحقول
        const fields = document.querySelectorAll('input:not([type="hidden"]), select, textarea');
        
        fields.forEach(field => {
            // تجاهل الحقول التي تحتوي بالفعل على أيقونة مساعدة
            if (field.parentElement?.querySelector('.help-icon')) {
                return;
            }
            
            const fieldId = field.id || field.name;
            const fieldLabel = this.getFieldLabel(field);
            
            // إضافة أيقونة مساعدة
            const icon = document.createElement('i');
            icon.className = 'fas fa-question-circle help-icon';
            icon.dataset.fieldId = fieldId;
            icon.dataset.fieldLabel = fieldLabel;
            icon.title = 'اضغط للمساعدة (أو F1)';
            
            // إدراج الأيقونة
            if (field.parentElement) {
                const wrapper = field.parentElement;
                
                // إذا كان الحقل داخل div.form-group
                if (wrapper.classList.contains('form-group') || wrapper.classList.contains('mb-3')) {
                    const label = wrapper.querySelector('label');
                    if (label && !label.querySelector('.help-icon')) {
                        label.appendChild(icon);
                    }
                } else {
                    field.insertAdjacentElement('afterend', icon);
                }
            }
        });
        
        console.log('Help icons added to', fields.length, 'fields');
    }
    
    getFieldLabel(field) {
        // البحث عن label المرتبط
        let label = null;
        
        if (field.id) {
            label = document.querySelector(`label[for="${field.id}"]`);
        }
        
        if (!label && field.parentElement) {
            label = field.parentElement.querySelector('label');
        }
        
        return label ? label.textContent.trim() : field.name;
    }
    
    addFloatingButton() {
        // إنشاء زر المساعدة العائم
        const button = document.createElement('button');
        button.className = 'floating-help-button';
        button.innerHTML = '<i class="fas fa-question"></i>';
        button.title = 'المساعدة (F1)';
        // تأكيد الموقع بعيداً عن زر الشات حتى لو وُجد CSS آخر
        button.style.top = '80px';
        button.style.left = '20px';
        button.style.bottom = 'auto';
        button.style.right = 'auto';
        button.onclick = () => this.showGeneralHelp();
        
        document.body.appendChild(button);
    }
    
    bindEvents() {
        // النقر على أيقونة المساعدة
        document.addEventListener('click', (e) => {
            if (e.target.classList.contains('help-icon') || 
                e.target.parentElement?.classList.contains('help-icon')) {
                const icon = e.target.classList.contains('help-icon') ? e.target : e.target.parentElement;
                const fieldId = icon.dataset.fieldId;
                const fieldLabel = icon.dataset.fieldLabel;
                this.showHelp(fieldId, fieldLabel);
            }
        });
        
        // F1 للمساعدة السياقية
        document.addEventListener('keydown', (e) => {
            if (e.key === 'F1') {
                e.preventDefault();
                const activeElement = document.activeElement;
                
                if (activeElement && (activeElement.tagName === 'INPUT' || 
                    activeElement.tagName === 'SELECT' || 
                    activeElement.tagName === 'TEXTAREA')) {
                    const fieldId = activeElement.id || activeElement.name;
                    const fieldLabel = this.getFieldLabel(activeElement);
                    this.showHelp(fieldId, fieldLabel);
                } else {
                    this.showGeneralHelp();
                }
            }
        });
        
        // Tooltip عند hover
        document.addEventListener('mouseover', (e) => {
            if (e.target.classList.contains('help-icon')) {
                const fieldId = e.target.dataset.fieldId;
                if (this.helpData[fieldId]) {
                    e.target.title = this.helpData[fieldId].title;
                }
            }
        });
    }
    
    showHelp(fieldId, fieldLabel) {
        console.log('Showing help for:', fieldId);
        
        let help = this.helpData[fieldId];
        
        // إذا لم تكن هناك مساعدة محددة، استخدم مساعدة عامة
        if (!help) {
            help = {
                title: fieldLabel || 'مساعدة',
                content: 'هذا الحقل يحتاج إلى معلومات صحيحة. يرجى التأكد من البيانات المدخلة.',
                tips: ['استخدم Tab للانتقال للحقل التالي', 'البيانات المطلوبة مميزة بـ *']
            };
        }
        
        let content = `
            <div class="contextual-help">
                <h5><i class="fas fa-info-circle"></i> ${help.title}</h5>
                <p>${help.content}</p>
        `;
        
        if (help.tips && help.tips.length > 0) {
            content += '<div class="help-tips"><strong><i class="fas fa-lightbulb"></i> نصائح مفيدة:</strong><ul>';
            help.tips.forEach(tip => {
                content += `<li>${tip}</li>`;
            });
            content += '</ul></div>';
        }
        
        if (help.html) {
            content += `<div class="help-additional">${help.html}</div>`;
        }
        
        if (help.video_url) {
            content += `
                <div class="help-video">
                    <iframe src="${help.video_url}" allowfullscreen></iframe>
                </div>
            `;
        }
        
        content += '</div>';
        
        this.showModal('المساعدة - ' + help.title, content);
    }
    
    showGeneralHelp() {
        const content = `
            <div class="general-help">
                <h5><i class="fas fa-book"></i> المساعدة العامة</h5>
                
                <h6>اختصارات لوحة المفاتيح:</h6>
                <table class="table table-sm table-bordered">
                    <thead>
                        <tr>
                            <th style="width: 30%">الاختصار</th>
                            <th>الوظيفة</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td><kbd>F1</kbd></td><td>عرض المساعدة السياقية</td></tr>
                        <tr><td><kbd>F3</kbd></td><td>البحث في الحسابات</td></tr>
                        <tr><td><kbd>F4</kbd></td><td>الآلة الحاسبة</td></tr>
                        <tr><td><kbd>F5</kbd></td><td>فاتورة جديدة</td></tr>
                        <tr><td><kbd>Ctrl + S</kbd></td><td>حفظ</td></tr>
                        <tr><td><kbd>Ctrl + F</kbd></td><td>بحث عام</td></tr>
                        <tr><td><kbd>Ctrl + P</kbd></td><td>طباعة</td></tr>
                        <tr><td><kbd>Tab</kbd></td><td>الانتقال للحقل التالي</td></tr>
                        <tr><td><kbd>Shift + Tab</kbd></td><td>العودة للحقل السابق</td></tr>
                        <tr><td><kbd>Esc</kbd></td><td>إلغاء / إغلاق</td></tr>
                    </tbody>
                </table>
                
                <div class="inline-help">
                    <i class="fas fa-info-circle"></i>
                    <strong>نصيحة:</strong> للحصول على مساعدة حول حقل معين، ضع المؤشر عليه واضغط <kbd>F1</kbd> أو انقر على أيقونة المساعدة 
                    <i class="fas fa-question-circle text-primary"></i>
                </div>
                
                <h6>الأسئلة الشائعة:</h6>
                <div class="faq-section">
                    <div class="faq-item">
                        <div class="faq-question" onclick="this.classList.toggle('active'); this.nextElementSibling.classList.toggle('show')">
                            <span>كيف أقوم بإنشاء فاتورة جديدة؟</span>
                            <i class="fas fa-chevron-down"></i>
                        </div>
                        <div class="faq-answer">
                            اضغط على "فواتير" من القائمة الجانبية، ثم "فاتورة جديدة"، أو اضغط <kbd>F5</kbd> من أي مكان.
                        </div>
                    </div>
                    
                    <div class="faq-item">
                        <div class="faq-question" onclick="this.classList.toggle('active'); this.nextElementSibling.classList.toggle('show')">
                            <span>كيف أبحث عن عميل؟</span>
                            <i class="fas fa-chevron-down"></i>
                        </div>
                        <div class="faq-answer">
                            في حقل العميل، ابدأ بكتابة الاسم أو رقم الهاتف، أو اضغط <kbd>F7</kbd> لفتح نافذة البحث المتقدم.
                        </div>
                    </div>
                    
                    <div class="faq-item">
                        <div class="faq-question" onclick="this.classList.toggle('active'); this.nextElementSibling.classList.toggle('show')">
                            <span>كيف أضيف خصماً على الفاتورة؟</span>
                            <i class="fas fa-chevron-down"></i>
                        </div>
                        <div class="faq-answer">
                            يمكنك إضافة خصم على مستوى البند أو على الإجمالي. اضغط <kbd>F8</kbd> لفتح آلة حاسبة الخصم.
                        </div>
                    </div>
                </div>
            </div>
        `;
        
        this.showModal('المساعدة العامة', content);
    }
    
    showModal(title, content) {
        // إنشاء modal إذا لم يكن موجوداً
        let modal = document.getElementById('helpModal');
        
        if (!modal) {
            modal = document.createElement('div');
            modal.id = 'helpModal';
            modal.className = 'modal fade';
            modal.setAttribute('tabindex', '-1');
            modal.innerHTML = `
                <div class="modal-dialog modal-lg modal-dialog-scrollable">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title" id="helpModalTitle"></h5>
                            <button type="button" class="btn-close btn-close-white" id="helpModalCloseX" aria-label="إغلاق"></button>
                        </div>
                        <div class="modal-body" id="helpModalBody"></div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-secondary" id="helpModalCloseBtn">
                                <i class="fas fa-times"></i> إغلاق
                            </button>
                        </div>
                    </div>
                </div>
            `;
            document.body.appendChild(modal);
            
            // ربط أحداث الإغلاق مباشرة
            document.getElementById('helpModalCloseX').onclick = () => this.closeModal();
            document.getElementById('helpModalCloseBtn').onclick = () => this.closeModal();
            
            // إغلاق بالنقر خارج النافذة
            modal.addEventListener('click', (e) => {
                if (e.target === modal) {
                    this.closeModal();
                }
            });
        }
        
        document.getElementById('helpModalTitle').innerHTML = title;
        document.getElementById('helpModalBody').innerHTML = content;
        
        // عرض modal - Simple direct approach
        modal.style.display = 'block';
        modal.classList.add('show');
        document.body.classList.add('modal-open');
        
        // إضافة backdrop
        let backdrop = document.getElementById('helpModalBackdrop');
        if (!backdrop) {
            backdrop = document.createElement('div');
            backdrop.id = 'helpModalBackdrop';
            backdrop.className = 'modal-backdrop fade show';
            backdrop.style.zIndex = '1040';
            backdrop.onclick = () => this.closeModal();
            document.body.appendChild(backdrop);
        } else {
            backdrop.style.display = 'block';
        }
    }
    
    closeModal() {
        const modal = document.getElementById('helpModal');
        const backdrop = document.getElementById('helpModalBackdrop');
        
        if (modal) {
            modal.style.display = 'none';
            modal.classList.remove('show');
        }
        if (backdrop) {
            backdrop.style.display = 'none';
        }
        document.body.classList.remove('modal-open');
    }
}

// تفعيل عند تحميل الصفحة
document.addEventListener('DOMContentLoaded', () => {
    window.contextualHelp = new ContextualHelp();
    console.log('Contextual Help ready!');
});

File: core/contextual_help.py
"""
نظام المساعدة السياقية
Contextual Help System
"""

from django.db import models


class HelpContent(models.Model):
    """محتوى المساعدة"""
    
    CONTEXT_TYPES = [
        ('form', 'نموذج'),
        ('field', 'حقل'),
        ('page', 'صفحة'),
        ('action', 'إجراء'),
        ('error', 'خطأ'),
    ]
    
    context_type = models.CharField(max_length=20, choices=CONTEXT_TYPES, verbose_name='نوع السياق')
    context_id = models.CharField(max_length=100, verbose_name='معرّف السياق')
    title = models.CharField(max_length=200, verbose_name='العنوان')
    content = models.TextField(verbose_name='المحتوى')
    content_html = models.TextField(blank=True, verbose_name='محتوى HTML')
    video_url = models.URLField(blank=True, verbose_name='رابط الفيديو')
    order = models.IntegerField(default=0, verbose_name='الترتيب')
    is_active = models.BooleanField(default=True, verbose_name='نشط')
    
    class Meta:
        verbose_name = 'محتوى مساعدة'
        verbose_name_plural = 'محتويات المساعدة'
        ordering = ['order', 'title']
        unique_together = ['context_type', 'context_id']
    
    def __str__(self):
        return f"{self.get_context_type_display()} - {self.title}"


class ContextualHelpSystem:
    """نظام المساعدة السياقية"""
    
    # مساعدات مدمجة للحقول الشائعة
    BUILT_IN_HELPS = {
        # حقول المبيعات
        'invoice_customer': {
            'title': 'اختيار العميل',
            'content': 'اختر العميل من القائمة أو ابحث عنه. يمكنك إضافة عميل جديد بالضغط على زر "+".',
            'tips': [
                'استخدم Ctrl+F للبحث السريع',
                'العملاء الأكثر تعاملاً يظهرون في الأعلى',
            ]
        },
        'invoice_items': {
            'title': 'إضافة منتجات',
            'content': 'أضف المنتجات للفاتورة. يمكنك استخدام الباركود أو البحث بالاسم.',
            'tips': [
                'اضغط F6 لإضافة منتج سريع',
                'استخدم Tab للانتقال بين الحقول',
                'الأسعار تُحدّث تلقائياً',
            ]
        },
        'invoice_discount': {
            'title': 'الخصم',
            'content': 'أدخل قيمة الخصم بالج.م. يمكنك تطبيق خصم على كل بند أو على الإجمالي.',
            'tips': [
                'اضغط F8 لفتح آلة حاسبة الخصم',
                'الخصم الإجمالي يوزع نسبياً على البنود',
            ]
        },
        
        # حقول المحاسبة
        'entry_account': {
            'title': 'اختيار الحساب',
            'content': 'اختر الحساب المناسب من دليل الحسابات. استخدم البحث للوصول السريع.',
            'tips': [
                'اضغط F3 للبحث في الحسابات',
                'الحسابات مرتبة حسب النوع',
                'يمكنك عرض رصيد الحساب قبل الاختيار',
            ]
        },
        'entry_debit_credit': {
            'title': 'المدين والدائن',
            'content': 'أدخل المبلغ في عمود المدين أو الدائن. يجب أن يتساوى مجموع المدين مع الدائن.',
            'tips': [
                'استخدم Tab للانتقال بين الأعمدة',
                'الأرصدة تُحسب تلقائياً',
                'تحذير إذا لم يتوازن القيد',
            ]
        },
        
        # حقول المخزون
        'product_min_stock': {
            'title': 'الحد الأدنى للمخزون',
            'content': 'حدد الحد الأدنى الذي يجب تنبيهك عند الوصول إليه.',
            'tips': [
                'يُفضل أن يكون كافٍ لأسبوع واحد',
                'سيصلك تنبيه عند النفاذ',
            ]
        },
        'product_reorder_quantity': {
            'title': 'كمية إعادة الطلب',
            'content': 'الكمية التي يُنصح بطلبها عند الوصول للحد الأدنى.',
            'tips': [
                'احسبها بناءً على معدل الاستهلاك',
                'خذ في الاعتبار وقت التوريد',
            ]
        },
        
        # حقول المشتريات
        'purchase_expected_delivery': {
            'title': 'تاريخ التسليم المتوقع',
            'content': 'حدد التاريخ المتوقع لاستلام البضاعة.',
            'tips': [
                'سيصلك تنبيه قبل الموعد بيوم',
                'تنبيه آخر إذا تأخر التسليم',
            ]
        },
        'purchase_payment_terms': {
            'title': 'شروط الدفع',
            'content': 'حدد شروط الدفع المتفق عليها مع المورد.',
            'tips': [
                'مثال: نقدي، آجل 30 يوم، تقسيط',
                'تؤثر على تقارير الذمم',
            ]
        },
    }
    
    @classmethod
    def get_help(cls, context_type, context_id):
        """الحصول على المساعدة لسياق معين"""
        # البحث في قاعدة البيانات
        try:
            db_help = HelpContent.objects.get(
                context_type=context_type,
                context_id=context_id,
                is_active=True
            )
            return {
                'title': db_help.title,
                'content': db_help.content,
                'content_html': db_help.content_html,
                'video_url': db_help.video_url,
                'source': 'database'
            }
        except HelpContent.DoesNotExist:
            pass
        
        # البحث في المساعدات المدمجة
        if context_id in cls.BUILT_IN_HELPS:
            return {
                **cls.BUILT_IN_HELPS[context_id],
                'source': 'built-in'
            }
        
        return None
    
    @classmethod
    def get_field_help(cls, field_name):
        """مساعدة خاصة بحقل"""
        return cls.get_help('field', field_name)
    
    @classmethod
    def get_form_help(cls, form_name):
        """مساعدة خاصة بنموذج"""
        return cls.get_help('form', form_name)
    
    @classmethod
    def get_page_help(cls, page_name):
        """مساعدة خاصة بصفحة"""
        return cls.get_help('page', page_name)
    
    @classmethod
    def create_help_tooltips(cls):
        """إنشاء tooltips لجميع الحقول"""
        tooltips = {}
        
        for field_id, help_data in cls.BUILT_IN_HELPS.items():
            tooltips[field_id] = {
                'title': help_data['title'],
                'content': help_data['content'],
                'tips': help_data.get('tips', [])
            }
        
        # إضافة من قاعدة البيانات
        db_helps = HelpContent.objects.filter(
            context_type='field',
            is_active=True
        )
        
        for help_item in db_helps:
            tooltips[help_item.context_id] = {
                'title': help_item.title,
                'content': help_item.content,
                'html': help_item.content_html
            }
        
        return tooltips


# JavaScript للمساعدة السياقية
CONTEXTUAL_HELP_JS = """
class ContextualHelp {
    constructor() {
        this.helpData = {};
        this.init();
    }
    
    init() {
        // تحميل بيانات المساعدة
        this.loadHelpData();
        
        // إضافة أيقونات المساعدة
        this.addHelpIcons();
        
        // معالجة الأحداث
        this.bindEvents();
    }
    
    loadHelpData() {
        // تحميل من الخادم
        fetch('/api/help/tooltips/')
            .then(response => response.json())
            .then(data => {
                this.helpData = data;
            });
    }
    
    addHelpIcons() {
        // إضافة أيقونة مساعدة لكل حقل له مساعدة
        document.querySelectorAll('input, select, textarea').forEach(field => {
            const fieldId = field.id || field.name;
            
            if (this.helpData[fieldId]) {
                const icon = document.createElement('i');
                icon.className = 'fas fa-question-circle help-icon';
                icon.dataset.fieldId = fieldId;
                icon.style.cursor = 'pointer';
                icon.style.marginLeft = '5px';
                icon.style.color = '#007bff';
                
                // إضافة بعد الحقل
                if (field.parentElement) {
                    field.parentElement.insertBefore(icon, field.nextSibling);
                }
            }
        });
    }
    
    bindEvents() {
        // عند النقر على أيقونة المساعدة
        document.addEventListener('click', (e) => {
            if (e.target.classList.contains('help-icon')) {
                const fieldId = e.target.dataset.fieldId;
                this.showHelp(fieldId);
            }
        });
        
        // عند التركيز على حقل (F1 للمساعدة)
        document.addEventListener('keydown', (e) => {
            if (e.key === 'F1') {
                e.preventDefault();
                const activeElement = document.activeElement;
                const fieldId = activeElement.id || activeElement.name;
                
                if (this.helpData[fieldId]) {
                    this.showHelp(fieldId);
                } else {
                    this.showGeneralHelp();
                }
            }
        });
    }
    
    showHelp(fieldId) {
        const help = this.helpData[fieldId];
        
        if (!help) return;
        
        let content = `
            <div class="contextual-help">
                <h5>${help.title}</h5>
                <p>${help.content}</p>
        `;
        
        if (help.tips && help.tips.length > 0) {
            content += '<div class="help-tips"><strong>نصائح:</strong><ul>';
            help.tips.forEach(tip => {
                content += `<li>${tip}</li>`;
            });
            content += '</ul></div>';
        }
        
        if (help.html) {
            content += help.html;
        }
        
        content += '</div>';
        
        // عرض في modal
        this.showModal('المساعدة', content);
    }
    
    showGeneralHelp() {
        const content = `
            <div class="general-help">
                <h5>مساعدة عامة</h5>
                <h6>اختصارات لوحة المفاتيح:</h6>
                <table class="table table-sm">
                    <tr><td><kbd>F1</kbd></td><td>المساعدة</td></tr>
                    <tr><td><kbd>Ctrl+S</kbd></td><td>حفظ</td></tr>
                    <tr><td><kbd>Ctrl+F</kbd></td><td>بحث</td></tr>
                    <tr><td><kbd>Tab</kbd></td><td>الحقل التالي</td></tr>
                    <tr><td><kbd>Esc</kbd></td><td>إلغاء</td></tr>
                </table>
                
                <p>للحصول على مساعدة حول حقل معين، ضع المؤشر عليه واضغط F1</p>
            </div>
        `;
        
        this.showModal('المساعدة العامة', content);
    }
    
    showModal(title, content) {
        // إنشاء modal إذا لم يكن موجوداً
        let modal = document.getElementById('helpModal');
        
        if (!modal) {
            modal = document.createElement('div');
            modal.id = 'helpModal';
            modal.className = 'modal fade';
            modal.innerHTML = `
                <div class="modal-dialog modal-lg">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title" id="helpModalTitle"></h5>
                            <button type="button" class="close" data-dismiss="modal">
                                <span>&times;</span>
                            </button>
                        </div>
                        <div class="modal-body" id="helpModalBody"></div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-secondary" data-dismiss="modal">إغلاق</button>
                        </div>
                    </div>
                </div>
            `;
            document.body.appendChild(modal);
        }
        
        document.getElementById('helpModalTitle').textContent = title;
        document.getElementById('helpModalBody').innerHTML = content;
        
        $(modal).modal('show');
    }
}

// تفعيل عند تحميل الصفحة
document.addEventListener('DOMContentLoaded', () => {
    window.contextualHelp = new ContextualHelp();
});
"""
